### ISSUE TITLE
Production build missing fontawesome imports when using dynamic imports with React.lazy

### ISSUE THREAD

**Issue by lindskogen** [vitejs/vite-plugin-react#32]
### Describe the bug

Font awesome icons shared across chunk-split boundaries are missing from every import except the first one when importing a font-awesome icon by path like so:

```tsx

// A.tsx
import { faPrint } from '@fortawesome/free-solid-svg-icons/faPrint';
```

```tsx

// B.tsx
import { faPrint } from '@fortawesome/free-solid-svg-icons/faPrint';
```

```tsx

// App.tsx
const A = lazy(() => import("./A"));
const B = lazy(() => import("./B"));

```

`faPrint` will only get bundled inside `A.xxxx.js` and is missing from `B.xxx.js`.

This causes the icons to disappear until you triggered the loading of `<B/>`'s chunk. For example, when using react-router, navigating back and forth between a page which renders `<A/>`. 

I've also managed to make a small repo that just uses `useState`. The icon won't be displayed until you press the button.

This issue doesn't happen if I import the icon from: `@fortawesome/free-solid-svg-icons` (the top level of the export).

Also, if I switch the order of the dynamic imports the issue doesn't happen:

```tsx

// App.tsx
const B = lazy(() => import("./B"));
const A = lazy(() => import("./A"));

```

(To reproduce, run `npm build` and then `npx serve -s dist`)

### Reproduction

https://stackblitz.com/edit/vitejs-vite-nkjxsu

### System Info

```shell
System:
    OS: macOS 12.5.1
    CPU: (10) arm64 Apple M1 Max
    Memory: 1.22 GB / 32.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.17.0 - /opt/homebrew/opt/node@16/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 8.15.0 - /opt/homebrew/opt/node@16/bin/npm
    Watchman: 2022.08.29.00 - /opt/homebrew/bin/watchman
  Browsers:
    Firefox: 104.0.2
    Safari: 15.6.1
  npmPackages:
    @vitejs/plugin-react: 2.1.0 => 2.1.0
    vite: 3.1.0 => 3.1.0
```

### Used Package Manager

yarn

### Logs

<details>
<summary>Click to expand!</summary>

```shell
vite:config bundled config file loaded in 559.10ms +0ms
  vite:esbuild init tsconfck (root: /home/projects/vitejs-vite-nkjxsu) +0ms
  vite:esbuild init tsconfck (root: /home/projects/vitejs-vite-nkjxsu) +1ms
  vite:esbuild init tsconfck (root: /home/projects/vitejs-vite-nkjxsu) +0ms
  vite:esbuild init tsconfck (root: /home/projects/vitejs-vite-nkjxsu) +1ms
  vite:esbuild init tsconfck end +1ms
  vite:esbuild init tsconfck end +0ms
  vite:esbuild init tsconfck end +0ms
  vite:esbuild init tsconfck end +0ms
  vite:config using resolved config: {
  vite:config   plugins: [
  vite:config     'vite:build-metadata',
  vite:config     'vite:pre-alias',
  vite:config     'alias',
  vite:config     'vite:react-babel',
  vite:config     'vite:react-refresh',
  vite:config     'vite:react-jsx',
  vite:config     'vite:modulepreload-polyfill',
  vite:config     'vite:resolve',
  vite:config     'vite:html-inline-proxy',
  vite:config     'vite:css',
  vite:config     'vite:esbuild',
  vite:config     'vite:json',
  vite:config     'vite:wasm-helper',
  vite:config     'vite:worker',
  vite:config     'vite:asset',
  vite:config     'vite:wasm-fallback',
  vite:config     'vite:define',
  vite:config     'vite:css-post',
  vite:config     'vite:build-html',
  vite:config     'vite:worker-import-meta-url',
  vite:config     'vite:force-systemjs-wrap-complete',
  vite:config     'vite:watch-package-data',
  vite:config     'commonjs',
  vite:config     'vite:data-uri',
  vite:config     'vite:asset-import-meta-url',
  vite:config     'vite:dynamic-import-vars',
  vite:config     'vite:import-glob',
  vite:config     'vite:build-import-analysis',
  vite:config     'vite:esbuild-transpile',
  vite:config     'vite:terser',
  vite:config     'vite:reporter',
  vite:config     'vite:load-fallback'
  vite:config   ],
  vite:config   optimizeDeps: {
  vite:config     disabled: 'build',
  vite:config     force: undefined,
  vite:config     include: [ 'react/jsx-dev-runtime' ],
  vite:config     esbuildOptions: { preserveSymlinks: undefined }
  vite:config   },
  vite:config   build: {
  vite:config     target: [ 'es2020', 'edge88', 'firefox78', 'chrome87', 'safari13' ],
  vite:config     polyfillModulePreload: true,
  vite:config     outDir: 'dist',
  vite:config     assetsDir: 'assets',
  vite:config     assetsInlineLimit: 4096,
  vite:config     cssCodeSplit: true,
  vite:config     cssTarget: [ 'es2020', 'edge88', 'firefox78', 'chrome87', 'safari13' ],
  vite:config     sourcemap: false,
  vite:config     rollupOptions: {},
  vite:config     minify: 'esbuild',
  vite:config     terserOptions: {},
  vite:config     write: true,
  vite:config     emptyOutDir: null,
  vite:config     manifest: false,
  vite:config     lib: false,
  vite:config     ssr: false,
  vite:config     ssrManifest: false,
  vite:config     reportCompressedSize: true,
  vite:config     chunkSizeWarningLimit: 500,
  vite:config     watch: null,
  vite:config     commonjsOptions: { include: [Array], extensions: [Array] },
  vite:config     dynamicImportVarsOptions: { warnOnError: true, exclude: [Array] }
  vite:config   },
  vite:config   resolve: { dedupe: [ 'react', 'react-dom' ], alias: [ [Object], [Object] ] },
  vite:config   configFile: '/home/projects/vitejs-vite-nkjxsu/vite.config.js',
  vite:config   configFileDependencies: [ '/home/projects/vitejs-vite-nkjxsu/vite.config.js' ],
  vite:config   inlineConfig: {
  vite:config     root: undefined,
  vite:config     base: undefined,
  vite:config     mode: undefined,
  vite:config     configFile: undefined,
  vite:config     logLevel: undefined,
  vite:config     clearScreen: undefined,
  vite:config     optimizeDeps: { force: undefined },
  vite:config     build: {}
  vite:config   },
  vite:config   root: '/home/projects/vitejs-vite-nkjxsu',
  vite:config   base: '/',
  vite:config   publicDir: '/home/projects/vitejs-vite-nkjxsu/public',
  vite:config   cacheDir: '/home/projects/vitejs-vite-nkjxsu/node_modules/.vite',
  vite:config   command: 'build',
  vite:config   mode: 'production',
  vite:config   ssr: {
  vite:config     format: 'esm',
  vite:config     target: 'node',
  vite:config     optimizeDeps: { disabled: true, esbuildOptions: [Object] }
  vite:config   },
  vite:config   isWorker: false,
  vite:config   mainConfig: null,
  vite:config   isProduction: true,
  vite:config   server: {
  vite:config     preTransformRequests: true,
  vite:config     middlewareMode: false,
  vite:config     fs: { strict: true, allow: [Array], deny: [Array] }
  vite:config   },
  vite:config   preview: {
  vite:config     port: undefined,
  vite:config     strictPort: undefined,
  vite:config     host: undefined,
  vite:config     https: undefined,
  vite:config     open: undefined,
  vite:config     proxy: undefined,
  vite:config     cors: undefined,
  vite:config     headers: undefined
  vite:config   },
  vite:config   env: { BASE_URL: '/', MODE: 'production', DEV: false, PROD: true },
  vite:config   assetsInclude: [Function: assetsInclude],
  vite:config   logger: {
  vite:config     hasWarned: false,
  vite:config     info: [Function: info],
  vite:config     warn: [Function: warn],
  vite:config     warnOnce: [Function: warnOnce],
  vite:config     error: [Function: error],
  vite:config     clearScreen: [Function: clearScreen],
  vite:config     hasErrorLogged: [Function: hasErrorLogged]
  vite:config   },
  vite:config   packageCache: Map(0) { set: [Function (anonymous)] },
  vite:config   createResolver: [Function: createResolver],
  vite:config   worker: {
  vite:config     format: 'iife',
  vite:config     plugins: [
  vite:config       'vite:build-metadata',
  vite:config       'vite:pre-alias',
  vite:config       'alias',
  vite:config       'vite:modulepreload-polyfill',
  vite:config       'vite:resolve',
  vite:config       'vite:html-inline-proxy',
  vite:config       'vite:css',
  vite:config       'vite:esbuild',
  vite:config       'vite:json',
  vite:config       'vite:wasm-helper',
  vite:config       'vite:worker',
  vite:config       'vite:asset',
  vite:config       'vite:wasm-fallback',
  vite:config       'vite:define',
  vite:config       'vite:css-post',
  vite:config       'vite:build-html',
  vite:config       'vite:worker-import-meta-url',
  vite:config       'vite:force-systemjs-wrap-complete',
  vite:config       'vite:watch-package-data',
  vite:config       'commonjs',
  vite:config       'vite:data-uri',
  vite:config       'vite:asset-import-meta-url',
  vite:config       'vite:dynamic-import-vars',
  vite:config       'vite:import-glob',
  vite:config       'vite:build-import-analysis',
  vite:config       'vite:esbuild-transpile',
  vite:config       'vite:terser',
  vite:config       'vite:reporter',
  vite:config       'vite:load-fallback'
  vite:config     ],
  vite:config     rollupOptions: {},
  vite:config     getSortedPlugins: [Function: getSortedPlugins],
  vite:config     getSortedPluginHooks: [Function: getSortedPluginHooks]
  vite:config   },
  vite:config   appType: 'spa',
  vite:config   experimental: { importGlobRestoreExtension: false, hmrPartialAccept: false },
  vite:config   getSortedPlugins: [Function: getSortedPlugins],
  vite:config   getSortedPluginHooks: [Function: getSortedPluginHooks]
  vite:config } +8ms
vite v3.1.0 building for production...
✓ 44 modules transformed.
dist/index.html                    0.38 KiB
dist/assets/A.8cf35845.js          0.84 KiB / gzip: 0.54 KiB
dist/assets/B.d3d76902.js          0.15 KiB / gzip: 0.14 KiB
dist/assets/index.es.02a05730.js   66.99 KiB / gzip: 19.01 KiB
dist/assets/index.6a88e9ed.js      140.47 KiB / gzip: 45.23 KiB
```
</details>

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to [vuejs/core](https://github.com/vuejs/core) instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by plamen89** [comment]
Same issue here. We are using multiple icons all over our application and now that we have migrated to vite some of them are missing. 
Another workaround that still lets you use destructuring is to use this syntax when importing the icons:

`import { definition as faTimes } from "@fortawesome/pro-light-svg-icons/faTimes"`

**Comment by templeman15** [comment]
I have this same problem as well but slightly different. I'm importing the icons I need into a single file. In that file I've been able to see some of the imports fail randomly, no single icon I can pin point it to. Across other files I'm importing `IconName` from font awesome but that is the only other font awesome import.

**Comment by mfusigh** [comment]
This happens without using lazy(). If an icon is imported in multiple files then they tend to be missing.

The "import { definition as ..." workaround only works if you reference that icon once. If you reference the same icon two times or more in separate files then they are missing again in the final build.

The icons will be missing in this case:

file A.jsx
`import { definition as faTimes } from "@fortawesome/pro-light-svg-icons/faTimes"`

file B.jsx
`import { definition as faTimes } from "@fortawesome/pro-light-svg-icons/faTimes"`


Both icons will appear in this case:

file A.jsx
`import { definition as faTimes } from "@fortawesome/pro-light-svg-icons/faTimes"`

file B.jsx (does not have "definition as")
`import { faTimes } from "@fortawesome/pro-light-svg-icons/faTimes"`

**Comment by Everduin94** [comment]
+1 to what @mfusigh mentioned. This happens without lazy. If you use a tree shakeable font awesome import in multiple files they will load momentarily and then disappear.

If you look at your network tab, you can see these individual icons that are imported multiple times and their response is (more-or-less) an empty object

![2023-01-19_09-01](https://user-images.githubusercontent.com/14320878/213476672-73e0f59a-d8ab-4a7e-8fca-9a6aab3cc0c8.png)

I never realized that this was only happening to icons used in multiple places until I read that comment. The workaround I used is to create a single file with the icons I use in multiple files so that I only have to import from font awesome in one file.

`icons.ts`
```typescript
import { definition as discord } from "@fortawesome/free-brands-svg-icons/faDiscord";
export const faDiscord =  discord
```

Then import from `icons.ts` instead of `@fortawesome` wherever you need to use it.

**Comment by ArnaudBarre** [comment]
Hi all,

I just looked at the issue and I don't think this is something specific with React. There is probably something in the FA structure that brake Rollup code spitting algorithm.

For information, esbuild correctly create a separated chunk for the shared icon (running `esbuild --bundle main.jsx --outdir=dist2 --format=esm --splitting --minify`) (Note: that's not a solution, because building a full Vite app with esbuild takes quite more than a command).

I have a lot of pending issue more specific to this plugin to look at, so I can't help you more, but you try to create a very small repo with only rollup and font awesome (using only `import()` and JS files without `React.lazy`) to isolate the issue and post it in the appropriate repo. Maybe the issue is not with rollup, but how Vite configures it or in the metadata of fortawesome libs.

**Comment by ilyary** [comment]
Having the same issue on Vue. Quick fix for rollup:
```
config.build.rollupOptions.output.manualChunks = (id) => {
    // FIX: Combine all FA assets into one chunk imported at an entry point
    if (/@fortawesome/.test(id)) {
        return 'fa';
    }
}
```

**Comment by xylish7** [comment]
@ArnaudBarre  I've encountered a similar issue using https://pictogrammers.com/library/mdi/. In development everything works ok, but in production I got the following error:
```
index-cb500412.js:40 Error: Minified React error #130; visit https://reactjs.org/docs/error-decoder.html?invariant=130&args[]=object&args[]= for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at Dr (index-cb500412.js:40:49054)
    at T (index-cb500412.js:38:17327)
    at ie (index-cb500412.js:40:1475)
    at ac (index-cb500412.js:40:45412)
    at ic (index-cb500412.js:40:39510)
    at Od (index-cb500412.js:40:39441)
    at il (index-cb500412.js:40:39300)
    at Go (index-cb500412.js:40:35718)
    at rc (index-cb500412.js:40:34674)
    at k (index-cb500412.js:25:1556)
index-cb500412.js:40 Error: Minified React error #130; visit https://reactjs.org/docs/error-decoder.html?invariant=130&args[]=object&args[]= for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at Dr (index-cb500412.js:40:49054)
    at T (index-cb500412.js:38:17327)
    at ie (index-cb500412.js:40:1475)
    at ac (index-cb500412.js:40:45412)
    at ic (index-cb500412.js:40:39510)
    at Od (index-cb500412.js:40:39441)
    at il (index-cb500412.js:40:39300)
    at Go (index-cb500412.js:40:35718)
    at rc (index-cb500412.js:40:34674)
    at k (index-cb500412.js:25:1556)
```

I've tried the same setup using `create-react-app` and everything works ok.
 
### Reproduce the issue:
1. Create a new react vite project
```
npm create vite@latest
```
2. Enter project name
3. Select `React` as a framework
4. Select `JavaScript` as a variant
5. Change directory to the create project
```
cd [PROJECT_NAME]
```
6. Add `@mdi/js` and `@mdi/react` packages.
```
npm i @mdi/js @mdi/react
```
7. Change the code in `App.jsx` with the following:
```jsx
import './App.css'
import Icon from '@mdi/react'
import { mdiPlus } from '@mdi/js'

function App() {
  return (
    <div className="App">
     <Icon path={mdiPlus} size={1}/>
    </div>
  )
}

export default App
```
8. Build the project
```
npm run build
```
9. Preview project
```
npm run preview
```
10. Go to `http://127.0.0.1:4173/`, inspect the code and you will get the error I've shown above.

**Comment by xylish7** [comment]
It seems like upgrading to `vite@4.3.3` fixed my issues.

**Comment by ArnaudBarre** [comment]
The issue persist for the fontawesome with latest version.
Made a smaller repro rollup specific: https://github.com/rollup/rollup/issues/5025

**Comment by ArnaudBarre** [comment]
This has been fixed upstream and available in rollup 3.24.1!
